python/rikai/spark/sql/codegen/fs.py

```diff
@@ -13,12 +13,121 @@
 #  limitations under the License.
 
 import secrets
-from typing import Dict
+from typing import Any, Dict, Optional, Union, IO, Mapping
 
+import yaml
+from jsonschema import ValidationError, validate
 from pyspark.sql import SparkSession
 
+from rikai.io import open_uri
 from rikai.logging import logger
 from rikai.spark.sql.codegen.base import Registry
+from rikai.spark.sql.exceptions import SpecError
+from rikai.spark.sql.schema import parse_schema
+
+__all__ = ["FileSystemRegistry"]
+
+# YAML-Spec SCHEMA
+SPEC_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "version": {
+            "type": "number",
+            "description": "Model SPEC format version",
+        },
+        "name": {"type": "string", "description": "Model name"},
+        "schema": {"type": "string"},
+        "model": {
+            "type": "object",
+            "description": "model description",
+            "properties": {
+                "uri": {"type": "string"},
+                "flavor": {"type": "string"},
+            },
+            "required": ["uri"],
+        },
+    },
+    "required": ["version", "schema", "model"],
+}
+
+
+class ModelSpec:
+    """Model Spec"""
+
+    def __init__(
+        self,
+        spec: Union[bytes, str, IO, Dict[str, Any]],
+        validate: bool = True,
+    ):
+        if not isinstance(spec, Mapping):
+            spec = yaml.load(spec, Loader=yaml.FullLoader)
+        self._spec = spec
+
+        if validate:
+            self.validate()
+
+    def validate(self):
+        """Validate model spec
+
+        Raises
+        ------
+        SpecError
+            If the spec is not well-formatted.
+        """
+        logger.debug("Validate spec: %s", self._spec)
+        try:
+            validate(instance=self._spec, schema=SPEC_SCHEMA)
+        except ValidationError as e:
+            raise SpecError(e.message) from e
+
+    @property
+    def version(self):
+        """Returns spec version."""
+        return self._spec["version"]
+
+    @property
+    def uri(self):
+        """Return Model URI"""
+        return self._spec["model"]["uri"]
+
+    @property
+    def flavor(self):
+        """Model flavor"""
+        return self._spec["model"].get("flavor", "")
+
+    @property
+    def schema(self):
+        """Return the output schema of the model."""
+        return parse_schema(self._spec["schema"])
+
+    @property
+    def options(self):
+        """Model options"""
+        return self._spec.get("options", {})
+
+
+def codegen_from_yaml(
+    spark: SparkSession,
+    uri: str,
+    name: Optional[str] = None,
+    options: Dict[str, str] = {},
+):
+    with open_uri(uri) as fobj:
+        spec = ModelSpec(fobj)
+
+    if spec.version != 1.0:
+        raise SpecError(
+            f"Only spec version 1.0 is supported, got {spec.version}"
+        )
+
+    if spec.flavor == "pytorch":
+        pass
```

Comment:
what's this?

Reply:
For every flavor, it will build a UDF to run inference differently.

Reply:
https://github.com/eto-ai/rikai/pull/119/files#diff-7fccc4b1dbc18bbb3a690fc62930ecddeb86c0f17dc274555a6a4affa61f7e8dR65